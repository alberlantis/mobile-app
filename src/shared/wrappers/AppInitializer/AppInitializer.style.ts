import { StyleSheet } from "react-native";

import colors from "src/theme/colors";

export default StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.BLACK,
  },
});
