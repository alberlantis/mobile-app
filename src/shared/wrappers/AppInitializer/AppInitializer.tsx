import React, { useEffect, useState, useCallback } from "react";
import { SafeAreaView, StatusBar, Pressable, Keyboard } from "react-native";

import { useSplash, useImageAssets } from "src/shared/hooks";
import s from "./AppInitializer.style";

interface ISplashProps {
  children: React.ReactNode;
}

const AppInitializer: React.FC<ISplashProps> = ({ children }) => {
  const [appIsReady, setAppIsReady] = useState(false);

  const { prepareSplash, hideSplash } = useSplash();
  const { preloadImagesAssets } = useImageAssets();

  useEffect(() => {
    async function initializeApp() {
      try {
        // Initialize all your assets/clients, fetch all your data here
        await prepareSplash();
        await preloadImagesAssets();
      } finally {
        setAppIsReady(true);
      }
    }
    initializeApp();
  }, [prepareSplash, preloadImagesAssets]);

  const onLayoutRootView = useCallback(async () => {
    if (appIsReady) {
      await hideSplash();
    }
  }, [appIsReady, hideSplash]);

  if (!appIsReady) {
    return null;
  }

  return (
    <SafeAreaView
      style={s.container}
      onLayout={onLayoutRootView}
      testID="app-initializer-container-id"
    >
      <StatusBar barStyle="light-content" />
      <Pressable
        style={s.container}
        accessible={false}
        onPress={Keyboard.dismiss}
      >
        {children}
      </Pressable>
    </SafeAreaView>
  );
};

export default AppInitializer;
