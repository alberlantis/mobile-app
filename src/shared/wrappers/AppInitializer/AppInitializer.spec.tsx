import React from "react";
import { render, waitFor, screen } from "@testing-library/react-native";
import { Text } from "react-native";
import AppInitializer from "./AppInitializer";

let resolveSplashPromise: (value?: boolean) => void;
const mockPrepareSplashPromise = new Promise(
  (resolve: (value?: boolean) => void) => {
    resolveSplashPromise = resolve;
  },
);
let resolvePreloadImageAssetsPromse: (value?: boolean) => void;
const mockPreloadImageAssetsPromise = new Promise(
  (resolve: (value?: boolean) => void) => {
    resolvePreloadImageAssetsPromse = resolve;
  },
);
const mockPreloadImageAssets = jest
  .fn()
  .mockResolvedValue(mockPreloadImageAssetsPromise);
const mockPrepareSplash = jest.fn().mockResolvedValue(mockPrepareSplashPromise);
const mockHidePlash = jest.fn();
jest.mock("src/shared/hooks", () => ({
  useSplash: jest.fn(() => ({
    prepareSplash: mockPrepareSplash,
    hideSplash: mockHidePlash,
  })),
  useImageAssets: jest.fn(() => ({
    preloadImagesAssets: mockPreloadImageAssets,
  })),
}));

describe("AppInitializer", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const renderComponent = () => {
    render(
      <AppInitializer>
        <Text>App</Text>
      </AppInitializer>,
    );
  };

  it("should render null initially and then should render children", async () => {
    renderComponent();

    expect(screen.toJSON()).toBeNull();
    expect(screen.toJSON()).toMatchSnapshot();

    await waitFor(async () => {
      resolvePreloadImageAssetsPromse();
      resolveSplashPromise();
    });

    await waitFor(() => {
      expect(screen.getByText("App")).toBeOnTheScreen();
      expect(screen.toJSON()).toMatchSnapshot();
    });
  });

  it("should prepare splash when initialize app", async () => {
    renderComponent();

    await waitFor(() => {
      expect(mockPrepareSplash).toHaveBeenCalledTimes(1);
    });
  });

  it("should prepare image assets when initialize app", async () => {
    renderComponent();

    await waitFor(() => {
      expect(mockPreloadImageAssets).toHaveBeenCalledTimes(1);
    });
  });

  it("should hide splash screen when all data is fetched and layout is ready", async () => {
    renderComponent();

    await waitFor(async () => {
      resolvePreloadImageAssetsPromse();
      resolveSplashPromise();
    });

    await waitFor(async () => {
      const safeAreaView = screen.getByTestId("app-initializer-container-id");
      safeAreaView.props.onLayout();

      await waitFor(() => {
        expect(mockHidePlash).toHaveBeenCalledTimes(1);
      });
    });
  });
});
