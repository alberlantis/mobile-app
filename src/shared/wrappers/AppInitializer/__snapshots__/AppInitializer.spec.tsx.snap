// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppInitializer should render null initially and then should render children 1`] = `null`;

exports[`AppInitializer should render null initially and then should render children 2`] = `
<RCTSafeAreaView
  onLayout={[Function]}
  style={
    {
      "backgroundColor": "#000",
      "flex": 1,
    }
  }
  testID="app-initializer-container-id"
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={false}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#000",
        "flex": 1,
      }
    }
  >
    <Text>
      App
    </Text>
  </View>
</RCTSafeAreaView>
`;
